### PR TITLE
[posix] optimize the source match table implementation

### PR DESCRIPTION
**Pull request by zhanglongxia**

### CODE REVIEW

File: examples/platforms/posix/openthread-core-posix-config.h [hunk 1]
```diff
@@ -136,4 +136,21 @@
...
 #define OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE 1
 #endif
+
+/**
+ * @def RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM
+ *
+ * The number of short source address table entries.
+ *
+ */
+#define RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM 32
```
Comment (bukepo) [inline examples/platforms/posix/openthread-core-posix-config.h]: #ifndef?

Comment (zhanglongxia) [reply examples/platforms/posix/openthread-core-posix-config.h]: Done

File: examples/platforms/posix/openthread-core-posix-config.h [hunk 2]
```diff
@@ -136,4 +136,25 @@
...
 #endif
+
+/**
+ * @def RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM
+ *
+ * The number of short source address table entries.
+ *
+ */
+#ifndef RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM
+#define RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM 10
```
Comment (bukepo) [inline examples/platforms/posix/openthread-core-posix-config.h]: Possible to define as `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN`?

Comment (zhanglongxia) [reply examples/platforms/posix/openthread-core-posix-config.h]: Updated.

File: examples/platforms/posix/radio.c [hunk 3]
```diff
@@ -240,7 +206,7 @@ static inline bool isDataRequestAndHasFramePending(const uint8_t *aFrame)
 
         if (sSrcMatchEnabled)
         {
-            hasFramePending = findShortAddress((uint16_t)(cur[1] << 8 | cur[0]));
+            hasFramePending = utilsSoftSrcMatchShortFindEntry((uint16_t)(cur[1] << 8 | cur[0])) < 0 ? false : true;
```
Comment (jwhui) [inline examples/platforms/posix/radio.c]: Replace with:
```c++
hasFramePending = utilsSoftSrcMatchShortFindEntry((uint16_t)(cur[1] << 8 | cur[0])) >= 0;
```
?

File: examples/platforms/posix/radio.c [hunk 4]
```diff
@@ -254,7 +220,7 @@ static inline bool isDataRequestAndHasFramePending(const uint8_t *aFrame)
 
         if (sSrcMatchEnabled)
         {
-            hasFramePending = findExtAddress((const otExtAddress *)cur);
+            hasFramePending = utilsSoftSrcMatchExtFindEntry((const otExtAddress *)cur) < 0 ? false : true;
```
Comment (jwhui) [inline examples/platforms/posix/radio.c]: Replace with:
```c++
hasFramePending = utilsSoftSrcMatchExtFindEntry((const otExtAddress *)cur) >= 0;
```
?

Comment (zhanglongxia) [reply examples/platforms/posix/radio.c]: Updated

File: examples/platforms/posix/radio.c (hunk 3 above)
Comment (zhanglongxia) [reply examples/platforms/posix/radio.c]: Updated.

**Comment by jwhui** [comment]
@zhanglongxia, please resolve conflicts.

**Comment by zhanglongxia** [comment]
@librasungirl I have updated the code, please review it again.

File: examples/platforms/utils/soft_source_match_table.c [hunk 5]
```diff
@@ -161,45 +161,45 @@ typedef struct srcMatchExtEntry
...
+    int16_t  entry    = -1;
     uint16_t checksum = sPanId;
 
     checksum += (uint16_t)aExtAddress->m8[0] | (uint16_t)(aExtAddress->m8[1] << 8);
     checksum += (uint16_t)aExtAddress->m8[2] | (uint16_t)(aExtAddress->m8[3] << 8);
     checksum += (uint16_t)aExtAddress->m8[4] | (uint16_t)(aExtAddress->m8[5] << 8);
     checksum += (uint16_t)aExtAddress->m8[6] | (uint16_t)(aExtAddress->m8[7] << 8);
 
-    for (uint8_t i = 0; i < RADIO_CONFIG_SRC_MATCH_EXT_ENTRY_NUM; i++)
+    for (uint16_t i = 0; i < RADIO_CONFIG_SRC_MATCH_EXT_ENTRY_NUM; i++)
```
Comment (librasungirl) [inline examples/platforms/utils/soft_source_match_table.c]: how about directly defining `i` as `int16_t`?

File: examples/platforms/utils/soft_source_match_table.h [hunk 6]
```diff
@@ -58,11 +58,11 @@ void utilsSoftSrcMatchSetPanId(uint16_t aPanId);
 #endif // RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM || RADIO_CONFIG_SRC_MATCH_EXT_ENTRY_NUM
 
 #if RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM
-int8_t utilsSoftSrcMatchShortFindEntry(const uint16_t aShortAddress);
+int16_t utilsSoftSrcMatchShortFindEntry(const uint16_t aShortAddress);
```
Comment (bukepo) [inline examples/platforms/utils/soft_source_match_table.h]: no need to use const

File: examples/platforms/utils/soft_source_match_table.c [hunk 7]
```diff
@@ -161,45 +161,45 @@ typedef struct srcMatchExtEntry
...
+            entry = (int16_t)i;
             break;
         }
     }
 
     return entry;
 }
 
-static inline void addToSrcMatchExtIndirect(uint8_t entry, const otExtAddress *aExtAddress)
+static inline void addToSrcMatchExtIndirect(uint16_t entry, const otExtAddress *aExtAddress)
```
Comment (bukepo) [inline examples/platforms/utils/soft_source_match_table.c]: remove inline?

File: examples/platforms/utils/soft_source_match_table.c [hunk 8]
```diff
@@ -212,7 +212,7 @@ static inline void addToSrcMatchExtIndirect(uint8_t entry, const otExtAddress *a
     srcMatchExtEntry[entry].allocated = true;
 }
 
-static inline void removeFromSrcMatchExtIndirect(uint8_t entry)
+static inline void removeFromSrcMatchExtIndirect(uint16_t entry)
```
Comment (bukepo) [inline examples/platforms/utils/soft_source_match_table.c]: remove inline?

File: examples/platforms/utils/soft_source_match_table.c [hunk 9]
```diff
@@ -161,45 +161,45 @@ typedef struct srcMatchExtEntry
...
+            entry = (int16_t)i;
             break;
         }
     }
 
     return entry;
 }
 
-static inline int8_t findSrcMatchExtAvailEntry(void)
+static inline int16_t findSrcMatchExtAvailEntry(void)
```
Comment (bukepo) [inline examples/platforms/utils/soft_source_match_table.c]: remove inline?

File: examples/platforms/utils/soft_source_match_table.c [hunk 10]
```diff
@@ -59,48 +59,48 @@ typedef struct srcMatchShortEntry
...
+            entry = (int16_t)i;
             break;
         }
     }
 
     return entry;
 }
 
-static inline int8_t findSrcMatchShortAvailEntry(void)
+static inline int16_t findSrcMatchShortAvailEntry(void)
```
Comment (bukepo) [inline examples/platforms/utils/soft_source_match_table.c]: remove inline?

File: examples/platforms/utils/soft_source_match_table.c (hunk 7 above)
Comment (zhanglongxia) [reply examples/platforms/utils/soft_source_match_table.c]: The [page](https://www.geeksforgeeks.org/inline-functions-cpp/) lists the cases that can't use `inline` keyword. This function doesn't match any item of the lists. I think it is ok to use `inline` in this function.

File: examples/platforms/utils/soft_source_match_table.c (hunk 8 above)
Comment (zhanglongxia) [reply examples/platforms/utils/soft_source_match_table.c]: The [page](https://www.geeksforgeeks.org/inline-functions-cpp/) lists the cases that can't use `inline` keyword. This function doesn't match any item of the lists. I think it is ok to use `inline` in this function.

File: examples/platforms/utils/soft_source_match_table.h (hunk 6 above)
Comment (zhanglongxia) [reply examples/platforms/utils/soft_source_match_table.h]: Updated.

File: tests/toranj/openthread-core-toranj-config.h [hunk 11]
```diff
@@ -392,6 +392,22 @@
  */
 #define OPENTHREAD_CONFIG_IP6_SLAAC_ENABLE    1
 
+/**
+ * @def RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM
+ *
+ * The number of short source address table entries.
+ *
+ */
+#define RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM OPENTHREAD_CONFIG_MLE_MAX_CHILDREN
```
Comment (bukepo) [inline tests/toranj/openthread-core-toranj-config.h]: `RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM` is not core config, maybe we should move these definitions into *examples/platforms/utils/soft_source_match_table.c*.

Comment (zhanglongxia) [reply tests/toranj/openthread-core-toranj-config.h]: Currently, all `example/platforms/utils` related configuration macros are defined in  `openthread-core-xxx-config.h`. So I defined the macro `RADIO_CONFIG_SRC_MATCH_SHORT_ENTRY_NUM` in  the core config file. I submitted an issue #4223 to discuss this topic. If we should move `example/platforms/utils` related configuration macros to other files. I will submit another PR to resolve this issue.

Comment (jwhui) [reply tests/toranj/openthread-core-toranj-config.h]: To address this specific issue, is it better to set the default value to `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` in: 
https://github.com/openthread/openthread/blob/cc91b2c72b6b1037302e662d90e7ed0ff674cab2/examples/platforms/utils/soft_source_match_table.h#L49
and

https://github.com/openthread/openthread/blob/cc91b2c72b6b1037302e662d90e7ed0ff674cab2/examples/platforms/utils/soft_source_match_table.h#L53

Comment (zhanglongxia) [reply tests/toranj/openthread-core-toranj-config.h]: Updated.

**Pull request merged**